ext/tunemygc/tunemygc_ext.c

```diff
@@ -72,21 +103,49 @@ static void tunemygc_gc_hook_i(VALUE tpval, void *data)
         case RUBY_INTERNAL_EVENT_GC_START:
             stat->stage = sym_gc_cycle_started;
             break;
+        case RUBY_INTERNAL_EVENT_GC_END_MARK:
+            stat->stage = sym_gc_cycle_mark_ended;
+            break;
         case RUBY_INTERNAL_EVENT_GC_END_SWEEP:
-            stat->stage = sym_gc_cycle_ended;
+            stat->stage = sym_gc_cycle_sweep_ended;
             break;
 #ifdef RUBY_INTERNAL_EVENT_GC_ENTER
         case RUBY_INTERNAL_EVENT_GC_ENTER:
             stat->stage = sym_gc_cycle_entered;
+            if (current_cycle != NULL) {
+                fprintf(stderr, "[TuneMyGc.ext] Reentrant GC Cycle?! Disabling!");
```

Comment:
As far as I know, no... but If it would the fsm would be broken and all data reported would be lies (in core that is), so I disable in case that happens. Which (currently?) it can't afaict and never did